rust-runtime/smithy-http/src/middleware.rs

```diff
@@ -17,9 +17,29 @@ use bytes::{Buf, Bytes};
 use http::Response;
 use http_body::Body;
 use std::error::Error;
+use std::future::Future;
+use std::pin::Pin;
 use tracing::trace;
 
 type BoxError = Box<dyn Error + Send + Sync>;
+type BoxFuture<T> = Pin<Box<dyn Future<Output = T> + Send>>;
+
+/// [`AsyncMapRequest`] defines an asynchronous middleware that transforms an [`operation::Request`].
+///
+/// Typically, these middleware will read configuration from the `PropertyBag` and use it to
+/// augment the request.
+///
+/// Most fundamental middleware is expressed as `AsyncMapRequest`'s synchronous cousin, `MapRequest`,
+/// including signing & endpoint resolution. `AsyncMapRequest` is used for async credential
+/// retrieval (e.g., from AWS STS's AssumeRole operation).
+pub trait AsyncMapRequest {
+    type Error: Into<BoxError> + 'static;
```

Comment:
I think we probably want `type Future` here instead of forcing a box future. If an implementor wants to use `BoxFuture` they can but this doesn't necessarily need boxing.

Reply:
If you want to punt for this pr, that's fine, just add a note that we can remove this in the future